src/NHibernate/Loader/DefaultEntityAliases.cs

```diff
@@ -10,48 +9,75 @@ namespace NHibernate.Loader
-		private readonly string suffixedDiscriminatorColumn;
-		private readonly string suffix;
-		private readonly string rowIdAlias;
-		private readonly IDictionary<string, string[]> userProvidedAliases;
+		private readonly string _suffix;
+		private readonly IDictionary<string, string[]> _userProvidedAliases;
+		private string _rowIdAlias;
 
 		public DefaultEntityAliases(ILoadable persister, string suffix)
-			: this(CollectionHelper.EmptyDictionary<string, string[]>(), persister, suffix) {}
+			: this(null, persister, suffix)
+		{
+		}
 
 		/// <summary>
-		/// Calculate and cache select-clause suffixes.
+		/// Calculate and cache select-clause aliases.
 		/// </summary>
 		public DefaultEntityAliases(IDictionary<string, string[]> userProvidedAliases, ILoadable persister, string suffix)
 		{
-			this.suffix = suffix;
-			this.userProvidedAliases = userProvidedAliases;
+			_suffix = suffix;
+			_userProvidedAliases = userProvidedAliases?.Count > 0 ? userProvidedAliases : null;
 
-			string[] keyColumnsCandidates = GetUserProvidedAliases(persister.IdentifierPropertyName, null);
-			if (keyColumnsCandidates == null)
-			{
-				suffixedKeyColumns = GetUserProvidedAliases(EntityPersister.EntityID, GetIdentifierAliases(persister, suffix));
-			}
-			else
-			{
-				suffixedKeyColumns = keyColumnsCandidates;
-			}
-			Intern(suffixedKeyColumns);
+			SuffixedKeyAliases = DetermineKeyAliases(persister);
+			SuffixedPropertyAliases = DeterminePropertyAliases(persister);
+			SuffixedDiscriminatorAlias = DetermineDiscriminatorAlias(persister);
+
+			SuffixedVersionAliases = persister.IsVersioned ? SuffixedPropertyAliases[persister.VersionProperty] : null;
+			//rowIdAlias is generated on demand in property
+		}
 
-			suffixedPropertyColumns = GetSuffixedPropertyAliases(persister);
-			suffixedDiscriminatorColumn =
-				GetUserProvidedAlias(AbstractEntityPersister.EntityClass, GetDiscriminatorAlias(persister, suffix));
-			if (persister.IsVersioned)
+		/// <summary>
+		/// Returns aliases for subclass persister
+		/// </summary>
+		public string[][] GetSuffixedPropertyAliases(ILoadable persister)
+		{
+			if (_userProvidedAliases == null)
+				return GetAllPropertyAliases(persister);
+
+			var propertyNames = persister.PropertyNames;
+			var suffixedPropertyAliases = new string[propertyNames.Length][];
+			for (var i = 0; i < propertyNames.Length; i++)
 			{
-				suffixedVersionColumn = suffixedPropertyColumns[persister.VersionProperty];
+				suffixedPropertyAliases[i] =
+					SafeGetUserProvidedAliases(propertyNames[i]) ??
+					GetPropertyAliases(persister, i);
 			}
-			else
+
+			return suffixedPropertyAliases;
+		}
+
+		public string[] SuffixedVersionAliases { get; }
+
+		public string[][] SuffixedPropertyAliases { get; }
+
+		public string SuffixedDiscriminatorAlias { get; }
+
+		public string[] SuffixedKeyAliases { get; }
+
+		// TODO: not visible to the user!
+		public string RowIdAlias => _rowIdAlias ?? (_rowIdAlias = Loadable.RowIdAlias + _suffix);
```

Comment:
I think none of us have realized we were changing a class which instances can be used by concurrent threads. In a concurrent usage situation, this micro-optimization can instead be a micro-performance loss! We would have better not have done it in my opinion.